mono/utils/mono-state.c

```diff
@@ -0,0 +1,476 @@
+/* Redefine MONO_ARCHITECTURE to include more information */
+#undef MONO_ARCHITECTURE
+#define MONO_ARCHITECTURE MONO_ARCH_ARCHITECTURE
+#endif
+
+static void
+mono_native_state_add_version (JsonWriter *writer)
+{
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "configuration");
+	mono_json_writer_object_begin(writer);
+
+	char *build = mono_get_runtime_callbacks ()->get_runtime_build_info ();
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "version");
+	mono_json_writer_printf (writer, "\"%s\",\n", build);
+
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "tlc");
+#ifdef HAVE_KW_THREAD
+	mono_json_writer_printf (writer, "\"__thread\",\n");
+#else
+	mono_json_writer_printf (writer, "\"normal\",\n");
+#endif /* HAVE_KW_THREAD */
+
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "sigsgev");
+#ifdef MONO_ARCH_SIGSEGV_ON_ALTSTACK
+	mono_json_writer_printf (writer, "\"altstack\",\n");
+#else
+	mono_json_writer_printf (writer, "\"normal\",\n");
+#endif
+
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "notifications");
+#ifdef HAVE_EPOLL
+	mono_json_writer_printf (writer, "\"epoll\",\n");
+#elif defined(HAVE_KQUEUE)
+	mono_json_writer_printf (writer, "\"kqueue\",\n");
+#else
+	mono_json_writer_printf (writer, "\"thread+polling\",\n");
+#endif
+
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "architecture");
+	mono_json_writer_printf (writer, "\"%s\",\n", MONO_ARCHITECTURE);
+
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "disabled_features");
+	mono_json_writer_printf (writer, "\"%s\",\n", DISABLED_FEATURES);
+
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "smallconfig");
+#ifdef MONO_SMALL_CONFIG
+	mono_json_writer_printf (writer, "\"enabled\",\n");
+#else
+	mono_json_writer_printf (writer, "\"disabled\",\n");
+#endif
+
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "bigarrays");
+#ifdef MONO_BIG_ARRAYS
+	mono_json_writer_printf (writer, "\"enabled\",\n");
+#else
+	mono_json_writer_printf (writer, "\"disabled\",\n");
+#endif
+
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "softdebug");
+#if !defined(DISABLE_SDB)
+	mono_json_writer_printf (writer, "\"enabled\",\n");
+#else
+	mono_json_writer_printf (writer, "\"disabled\",\n");
+#endif
+
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "interpreter");
+#ifndef DISABLE_INTERPRETER
+	mono_json_writer_printf (writer, "\"enabled\",\n");
+#else
+	mono_json_writer_printf (writer, "\"disabled\",\n");
+#endif
+
+	mono_json_writer_indent (writer);
+	mono_json_writer_object_key(writer, "llvm_support");
+#ifdef MONO_ARCH_LLVM_SUPPORTED
+#ifdef ENABLE_LLVM
+	mono_json_writer_printf (writer, "\"%s\"\n", LLVM_VERSION);
+#else
+	mono_json_writer_printf (writer, "\"disabled\"\n");
+#endif
+#endif
```

Comment:
how do we plan to keep this feature list in sync with the one in driver.c that we show during `mono --version`?

Reply:
If you want me to make both refer back to a common source of truth, can we address that in a future PR?

Reply:
Sure.